src/utils/ParseRouteUtils.js

```diff
@@ -532,8 +534,137 @@ function parseRoute(resp: Object, destinationName: string) {
+ */
+async function parseDetailedRoute(resp: Object, destinationName: string) {
+    const routeInfo = await parseRoute(resp, destinationName);
+
+    const {
+        departureTime,
+        arrivalTime,
+        directions,
+        startCoords,
+        endCoords,
+        boundingBox,
+        numberOfTransfers,
+    } = routeInfo;
+
+    let travelDistance = geolib.getDistance(
+        { latitude: startCoords.lat, longitude: startCoords.long },
+        { latitude: endCoords.lat, longitude: endCoords.long }
+    );
+
+    travelDistance *= METERS_TO_MILES;
+    routeInfo.travelDistance = travelDistance;
+
+    let totalDuration = new Date(arrivalTime) - new Date(departureTime);
+    totalDuration /= 1000;
+    totalDuration /= 60;
+    routeInfo.totalDuration = Math.round(totalDuration);
+
+    routeInfo.startName = routeInfo.directions[0].name;
+    routeInfo.endName = destinationName;
+
+    // routeSummary
+
+    /* routeSummaryElement {
+        stopName: String,
+        direction: RouteSummaryDirection?,
+        shouldStayOnBus: Bool,
+    }
+    */
+
+    /* routeSummaryDirection {
+        type: null || 'walk' || 'bus',
+        routeNumber: null || Int,
+    }
+    */
+
+    let routeSummary = [];
+
+    let walkOnlyRoute = true;
+    for (let i = 0; i < directions.length; i++) {
+        walkOnlyRoute = walkOnlyRoute && (directions[i].type === 'walk');
+    }
+
+    // if walk only route, the first element of routeSummary will be 'walk'
+    if (walkOnlyRoute) {
+        routeSummary.push({
+            stopName: directions[0].name,
+            direction: { type: 'walk', routeNumber: null },
+            shouldStayOnBus: false,
+        });
+        routeSummary.push({
+            stopName: directions[directions.length - 1].name,
+            direction: { type: null, routeNumber: null },
+            shouldStayOnBus: false,
+        });
+    } else {
+        let indexOfFirstBus = -1;
+        for (let i = 0; i < directions.length; i++) {
+            if (directions[i].type === 'depart') {
+                indexOfFirstBus = i;
+                break;
+            }
+        }
+
+        for (let i = indexOfFirstBus; i < directions.length; i++) {
+            let newRouteSummary = { stopName: directions[i].name, shouldStayOnBus: false };
+            switch (directions[i].type) {
+                case 'depart':
+                    newRouteSummary.direction = { type: 'bus', routeNumber: directions[i].routeNumber };
+                    break;
+                case 'walk':
+                    newRouteSummary.direction = { type: 'walk', routeNumber: null };
+                    break;
+                default:
+                    break;
+            }
+            routeSummary.push(newRouteSummary);
+        }
+
+        routeSummary.push({
+            stopName: directions[directions.length - 1].name,
+            direction: { type: null, routeNumber: null },
+            shouldStayOnBus: false,
+        });
```

Comment:
See condenseRoute in ParseRouteUtils, only one for-loop through the directions is done for all logic and parsing to simplify logic and save computation time. Therefore, place all this functionality within a single loop to avoid having to iterate through the directions more than once needlessly.